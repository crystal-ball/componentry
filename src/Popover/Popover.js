import activeContainer from '../active-container-factory'
import activeContent from '../active-content-factory'
import activeTrigger from '../active-trigger-factory'
import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

const Popover = activeContainer('popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
})
export default Popover

Popover.Body = function PopoverBody(props) {
  return elem({
    componentClassNames: 'popover-body',
    ...useTheme('PopoverBody'),
    ...props,
  })
}
Popover.Body.displayName = 'PopoverBody'

Popover.Content = activeContent('popover', {
  arias: { id: true, role: 'tooltip', hidden: true },
  popper: true,
})

Popover.Heading = function PopoverHeading(props) {
  return elem({
    as: 'h3',
    componentClassNames: 'popover-heading',
    ...useTheme('PopoverHeading'),
    ...props,
  })
}
Popover.Heading.displayName = 'PopoverHeading'

Popover.Trigger = activeTrigger('popover', {
  arias: { describedby: true },
})
