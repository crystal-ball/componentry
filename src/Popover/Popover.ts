import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionBuilder } from '../utils/active-action-component-builder'
import { activeContentBuilder } from '../utils/active-content-component-builder'
import { staticComponent } from '../utils/static-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../utils/types'

export interface PopoverProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface PopoverActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverBodyProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {}

export interface PopoverContentProps extends ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverHeadingProps extends ComponentBaseProps<'div'> {}

export interface Popover {
  (props: PopoverProps): React.ReactElement
  /**
   * [Popover action component üìù](https://componentry.design/components/popover)
   */
  Action: React.FC<PopoverActionProps>
  /**
   * [Popover body component üìù](https://componentry.design/components/popover)
   */
  Body: React.FC<PopoverBodyProps>
  /**
   * [Popover content component üìù](https://componentry.design/components/popover)
   */
  Content: React.FC<PopoverContentProps>
  /**
   * [Popover heading component üìù](https://componentry.design/components/popover)
   */
  Heading: React.FC<PopoverHeadingProps>
}

/**
 * [Popover component üìù](https://componentry.design/components/popover)
 * @experimental
 */
export const Popover = activeContainerBuilder<PopoverProps>('Popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
}) as Popover

Popover.Action = activeActionBuilder<PopoverActionProps>('PopoverAction', {
  aria: { describedby: true },
})

Popover.Body = staticComponent<PopoverBodyProps>('PopoverBody')

Popover.Content = activeContentBuilder<PopoverContentProps>('PopoverContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})

Popover.Heading = staticComponent<PopoverHeadingProps>('PopoverHeading', {
  as: 'h3',
})
