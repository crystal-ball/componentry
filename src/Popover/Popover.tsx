import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionComponent } from '../utils/active-action-component-builder'
import { activeContentComponent } from '../utils/active-content-component-builder'
import { staticComponent } from '../utils/static-component-builder'
import {
  BaseActiveActionProps,
  BaseActiveContainerProps,
  BaseActiveContentProps,
  BaseProps,
} from '../utils/base-types'

export interface PopoverProps
  extends BaseActiveContainerProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface PopoverActionProps
  extends BaseActiveActionProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverBodyProps
  extends BaseActiveContentProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface PopoverContentProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverHeadingProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface Popover {
  (props: PopoverProps): React.ReactElement
  /**
   * [Popover action component üìù](https://componentry.design/components/popover)
   */
  Action: React.FC<PopoverActionProps>
  /**
   * [Popover body component üìù](https://componentry.design/components/popover)
   */
  Body: React.FC<PopoverBodyProps>
  /**
   * [Popover content component üìù](https://componentry.design/components/popover)
   */
  Content: React.FC<PopoverContentProps>
  /**
   * [Popover heading component üìù](https://componentry.design/components/popover)
   */
  Heading: React.FC<PopoverHeadingProps>
}

/**
 * [Popover component üìù](https://componentry.design/components/popover)
 */
export const Popover = activeContainerBuilder<PopoverProps>('Popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
}) as Popover

Popover.Action = activeActionComponent('PopoverAction', {
  aria: { describedby: true },
})

Popover.Body = staticComponent<PopoverBodyProps>('PopoverBody', {
  componentCx: 'üÖ≤-popover-body',
})

Popover.Content = activeContentComponent<PopoverContentProps>('PopoverContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})

Popover.Heading = staticComponent<PopoverHeadingProps>('PopoverHeading', {
  as: 'h3',
  componentCx: 'üÖ≤-popover-heading',
})
