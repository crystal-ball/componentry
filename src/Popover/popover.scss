/* componentry/src/Popover/popover */

// Bootstrap popovers are inserted into the DOM at the end of the body on hover and
// the location is assigned with transforms. Componentry instead sets the popover
// parent container to position relative so that the absolute positioned content
// displays relative to parent. This means that Componentry has custom popover styles
// instead of using Bootstrap's

// Container Element
.popover {
  // Default align content to dead center of trigger, the specific content styles can
  // then easily move to top/left/right/bottom where needed
  align-items: center;
  display: inline-flex; // flex instead of block for easier center alignment of content
  justify-content: center;
  position: relative;

  // Directional styles

  &.top {
    .popover-content-container {
      bottom: 100%;
      justify-content: center;
    }

    .popover-content {
      // Top requires bumping bottom 100% again to compensate for container not having
      // any height
      bottom: 100%;
      margin-bottom: $popover-arrow-width;
    }

    .tip-container {
      top: 100%;
      width: 100%;
    }

    .tip:after {
      bottom: $popover-arrow-width * 0.5;
    }
  }

  &.right {
    .popover-content-container {
      left: 100%;
      align-items: center;
    }

    .popover-content {
      margin-left: $popover-arrow-width;
    }

    .tip-container {
      height: 100%;
      right: 100%;
    }

    .tip:after {
      left: $popover-arrow-width * 0.5;
    }
  }

  &.bottom {
    .popover-content-container {
      top: 100%;
      justify-content: center;
    }

    .popover-content {
      margin-top: $popover-arrow-width;
    }

    .tip-container {
      bottom: 100%;
      width: 100%;
    }

    .tip:after {
      top: $popover-arrow-width * 0.5;
    }
  }

  &.left {
    .popover-content-container {
      right: 100%;
      align-items: center;
    }

    .popover-content {
      margin-right: $popover-arrow-width;
      right: 0;
    }

    .tip-container {
      height: 100%;
      left: 100%;
    }

    .tip:after {
      right: $popover-arrow-width * 0.5;
    }
  }
}

// --- Popover trigger ------------------------------------

.popover-action-primary {
  @include active-trigger-element;
}

// Content container overrides the width constraints for parent element
.popover-content-container {
  width: $popover-max-width;
  position: absolute;
  display: inline-flex;
}

// Content Element - The content element should be used to wrap a body/header element
.popover-content {
  // Reset font properties for consistent font styling regardless of cascading values
  @include reset-text();
  display: block;
  position: absolute;
  max-width: $popover-max-width;
  z-index: $zindex-popover;

  background-clip: padding-box;
  background-color: $popover-bg;
  border: $popover-border-width solid $popover-border-color;
  font-size: $font-size-sm;
  // Allow breaking very long words so they don't overflow the popover's bounds
  word-wrap: break-word;

  box-shadow: box-shadow($popover-box-shadow);
  @include border-radius($border-radius-lg);

  .tip-container {
    align-items: center;
    display: flex;
    justify-content: center;
    overflow: hidden;
    position: absolute;
  }

  // The .tip container creates a containing box of the same height and width as the
  // tip with overflow hidden. Then inside that container we create a square of the
  // same dimensions, and rotate it 45 degrees to create the tip effect
  .tip {
    height: $popover-arrow-width;
    overflow: hidden;
    width: $popover-arrow-width;

    &:after {
      border: $popover-border-width solid $popover-border-color;
      // Do not use box-shadow mixin, output is wrapped in a check against
      // $enable-shadows, but we always want the arrow box-shadow included
      box-shadow: $popover-box-shadow;
      background: $popover-bg;
      content: '';
      height: $popover-arrow-width;
      position: absolute;
      transform: rotate(45deg);
      width: $popover-arrow-width;
    }
  }
}

// Body element should contain popover inner content
.popover-body {
  padding: $popover-body-padding-y $popover-body-padding-x;
  color: $popover-body-color;
}

// --------------------------------------------------------
// Heading element

.popover-heading {
  padding: $popover-header-padding-y $popover-header-padding-x;
  // Try to guarantee that spacing is correct regardless of div vs h* usage
  margin-top: 0;
  margin-bottom: 0;
  font-size: $font-size-base;
  color: $popover-header-color;
  background-color: $popover-header-bg;
  border-bottom: $popover-border-width solid darken($popover-header-bg, 5%);
  $offset-border-width: calc(#{$border-radius-lg} - #{$popover-border-width});
  @include border-top-radius($offset-border-width);

  &:empty {
    display: none;
  }
}
