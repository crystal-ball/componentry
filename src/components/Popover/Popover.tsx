import React from 'react'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
} from '../../utils/base-types'
import { staticComponent } from '../../utils/static-component-builder'
import { UtilityProps } from '../../utils/utility-classes'
import { Button } from '../Button/Button'

export interface PopoverProps
  extends ActiveContainerBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface PopoverActionProps
  extends ActiveActionBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverBodyProps
  extends ActiveContentBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface PopoverContentProps
  extends UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverHeadingProps
  extends UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {
  as?: React.ElementType
}

export interface Popover {
  (props: PopoverProps): React.ReactElement
  /**
   * [Popover action component üìù](https://componentry.design/components/popover)
   */
  Action: React.FC<PopoverActionProps>
  /**
   * [Popover body component üìù](https://componentry.design/components/popover)
   */
  Body: React.FC<PopoverBodyProps>
  /**
   * [Popover content component üìù](https://componentry.design/components/popover)
   */
  Content: React.FC<PopoverContentProps>
  /**
   * [Popover heading component üìù](https://componentry.design/components/popover)
   */
  Heading: React.FC<PopoverHeadingProps>
}

/**
 * [Popover component üìù](https://componentry.design/components/popover)
 * @experimental
 */
export const Popover = activeContainerBuilder<PopoverProps>('Popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
}) as Popover

Popover.Action = activeActionBuilder<PopoverActionProps>('PopoverAction', {
  aria: { describedby: true },
  defaultAs: Button,
})

function PopoverContentElement({
  children,
  // renderArrow = false, TODO
  ...rest
}: {
  children: React.ReactNode
  // renderArrow?: boolean
}) {
  return (
    <div {...rest}>
      {/* {renderArrow && <div className='C9Y-PopoverArrow' />} */}
      <div className='C9Y-PopoverContentContents'>{children}</div>
    </div>
  )
}

Popover.Content = activeContentBuilder<PopoverContentProps>('PopoverContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  defaultAs: PopoverContentElement,
})

Popover.Heading = staticComponent<PopoverHeadingProps>('PopoverHeading', {
  as: 'h3',
})

Popover.Body = staticComponent<PopoverBodyProps>('PopoverBody')
