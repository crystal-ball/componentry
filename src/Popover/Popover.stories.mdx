import { Meta, Title, Props } from '@storybook/addon-docs/blocks'
import { select } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import Popover from './Popover'
import { ElementsProvider } from '../story-resources'

<Meta title='Components|Popover' component={Popover} />

<ElementsProvider>

# Popovers

<Popover
  direction={select('direction', ['top', 'right', 'bottom', 'left'], 'right')}
  onActivate={action('onActivate event')}
  onActivated={action('onActivated event')}
  onDeactivate={action('onDeactivate event')}
  onDeactivated={action('onDeactivated event')}
>
  <Popover.Trigger>Toggle</Popover.Trigger>
  <Popover.Content>
    <Popover.Heading>Fun Fact!</Popover.Heading>
    <Popover.Body>
      <span>
        The new Texas Instrument calculators have ABC keyboards because if they had QWERTY
        keyboards, they would be considered computers and wouldnâ€™t be allowed for
        standardized test taking.
      </span>
      <hr />
      <img
        src='https://media.giphy.com/media/KVUhxQAqrAV3i/giphy.gif'
        className='img-fluid mw-100'
        alt="that's awesome"
      />
    </Popover.Body>
  </Popover.Content>
</Popover>

<Props of={Popover} />

</ElementsProvider>
