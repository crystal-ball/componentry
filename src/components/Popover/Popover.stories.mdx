import { Meta, Story } from '@storybook/addon-docs'

import { Popover } from './Popover'

<Meta title='Components/Popover' component={Popover} />

# Popover

Popovers are dialogs that may contain interactive elements.

<Story name='Popover'>
  <Popover active direction='bottom'>
    <Popover.Action>Toggle</Popover.Action>
    <Popover.Content>
      <Popover.Heading>Fun Fact!</Popover.Heading>
      <Popover.Body>
        The new Texas Instrument calculators have ABC keyboards because if they had QWERTY
        keyboards, they would be considered computers and wouldnâ€™t be allowed for
        standardized test taking.
      </Popover.Body>
    </Popover.Content>
  </Popover>
</Story>
