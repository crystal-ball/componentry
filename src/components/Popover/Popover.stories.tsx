import type { Meta, StoryObj } from '@storybook/react'

import { Popover } from './Popover'

const meta: Meta<typeof Popover> = {
  component: Popover,
}

export default meta
type Story = StoryObj<typeof Popover>

export const Primary: Story = {
  render: () => (
    <Popover active direction='bottom'>
      <Popover.Action>Toggle</Popover.Action>
      <Popover.Content>
        <Popover.Heading>Fun Fact!</Popover.Heading>
        <Popover.Body>
          The new Texas Instrument calculators have ABC keyboards because if they had
          QWERTY keyboards, they would be considered computers and wouldnâ€™t be allowed for
          standardized test taking.
        </Popover.Body>
      </Popover.Content>
    </Popover>
  ),
}
