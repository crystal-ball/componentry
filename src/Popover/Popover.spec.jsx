import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import Popover from './Popover'
// import activationTestSuite from '../utils-test/activation-tests'
import elementTests from '../utils-test/element-tests'
import setupActiveContext from '../utils-test/setup-active-context'

describe('<Popover />', () => {
  // Basic library activation test suite
  // TODO: Fix me
  // activationTestSuite(Popover)
  // Basic library element test suite
  elementTests(Popover)
  elementTests(Popover.Trigger, setupActiveContext())
  elementTests(Popover.Content, setupActiveContext())
  elementTests(Popover.Header)
  elementTests(Popover.Body)

  it('renders the correct directional classes using direction', () => {
    const wrapper = shallow(
      <Popover>
        <Popover.Trigger>Trigger</Popover.Trigger>
        <Popover.Content>Content</Popover.Content>
      </Popover>,
    )

    expect(wrapper.find('.popover.right').length).toEqual(1) // default
    wrapper.setProps({ direction: 'top' })
    expect(wrapper.find('.popover.top').length).toEqual(1)
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Drawer /> snapshots', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Popover>
          <Popover.Trigger>Trigger</Popover.Trigger>
          <Popover.Content>Content</Popover.Content>
        </Popover>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
