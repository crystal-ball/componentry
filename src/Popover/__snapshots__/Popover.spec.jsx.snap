// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> snapshots renders correctly 1`] = `
<div
  className="popover"
  data-id="guid"
  data-test="popover-container"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    aria-describedby="guid"
    className="popover-toggle btn"
    data-test="popover-trigger"
    onClick={[Function]}
    type="button"
    value={undefined}
  >
    Trigger
  </button>
  <div
    className="popover-content-container"
  >
    <div
      aria-hidden="true"
      className="popover-content"
      data-test="popover-content"
      id="guid"
      role="tooltip"
    >
      <div
        className="tip-container"
      >
        <div
          className="tip"
        />
      </div>
      Content
    </div>
  </div>
</div>
`;
